graftabledag/graftabledag.go

```diff
@@ -0,0 +1,151 @@
+package graftabledag
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/ipfs/go-cid"
+
+	"github.com/quorumcontrol/chaintree/chaintree"
+	"github.com/quorumcontrol/chaintree/dag"
+)
+
+type GraftableDag interface {
+	GlobalResolve(ctx context.Context, path chaintree.Path) (value interface{}, remaining chaintree.Path, err error)
+	OriginDag() *dag.Dag
+	DagGetter() DagGetter
+}
+
+type DagGetter interface {
+	GetTip(ctx context.Context, did string) (*cid.Cid, error)
+	GetLatest(ctx context.Context, did string) (*chaintree.ChainTree, error)
+}
+
+type GraftedDag struct {
+	dagCache  *lru.Cache // Is this premature optimization?
+	dagGetter DagGetter
+	origin    *dag.Dag
+}
+
+var _ GraftableDag = (*GraftedDag)(nil)
+
+func New(origin *dag.Dag, dagGetter DagGetter) (*GraftedDag, error) {
+	cache, err := lru.New(16)
+	if err != nil {
+		return nil, fmt.Errorf("could not create cache for GraftableDag: %w", err)
+	}
+
+	return &GraftedDag{
+		dagCache:  cache,
+		dagGetter: dagGetter,
+		origin:    origin,
+	}, nil
+}
+
+func (gd *GraftedDag) getChaintreeDag(ctx context.Context, did string) (*dag.Dag, error) {
+	tip, err := gd.dagGetter.GetTip(ctx, did)
+	if err != nil {
+		return nil, fmt.Errorf("could not get tip for %s: %w", did, err)
+	}
+
+	if uncastDag, ok := gd.dagCache.Get(tip); ok {
+		if ctDag, ok := uncastDag.(*dag.Dag); ok {
+			return ctDag, nil
+		}
+	}
+
+	chainTree, err := gd.dagGetter.GetLatest(ctx, did)
+	if err != nil {
+		return nil, fmt.Errorf("could not get latest for %s: %w", did, err)
+	}
+
+	gd.dagCache.Add(chainTree.Dag.Tip, chainTree.Dag)
+
+	return chainTree.Dag, nil
+}
+
+// TODO: Does this need loop detection? It's OK to go through the same chaintree more than
+//  once, just not the same chaintree+path.
+func (gd *GraftedDag) resolveRecursively(ctx context.Context, path chaintree.Path, d *dag.Dag) (value interface{}, remaining chaintree.Path, err error) {
+	value, remaining, err = d.Resolve(ctx, path)
+	if err != nil {
+		return value, remaining, err
+	}
+
+	didPaths := make([]chaintree.Path, 0)
+	values := make([]interface{}, 0)
+
+	switch v := value.(type) {
+	case string:
+		if strings.HasPrefix(v, "did:tupelo:") {
```

Comment:
not for now, but just noting that we could theoretically resolve *other* dids too.

Reply:
Would be very cool!